scipy/optimize/_numdiff.py

```diff
@@ -0,0 +1,297 @@
+    scheme : ndarray
+        Array containing 0 and 1. If 0 then use central finite difference,
+        if 1 then use 3-point forward finite difference.
+    """
+    lb, ub = bounds
+    if lb is None and ub is None:
+        return h, np.zeros_like(h, dtype=int)
+
+    if lb is None:
+        lb = np.empty_like(h)
+        lb.fill(-np.inf)
+
+    if ub is None:
+        ub = np.empty_like(h)
+        ub.fill(np.inf)
+
+    scheme = np.zeros_like(h, dtype=int)
+    h_adjusted = h.copy()
+
+    lower_dist = x0 - lb
+    upper_dist = ub - x0
+
+    central = (lower_dist >= h) & (upper_dist >= h)
+
+    forward = ~central & (upper_dist >= 2 * h)
+    scheme[forward] = 1
+    h_adjusted[forward] = h[forward]
+
+    backward = ~central & (lower_dist >= 2 * h)
+    scheme[backward] = 1
+    h_adjusted[backward] = -h[backward]
+
+    central_adjusted = ~central & ~backward & ~forward
+    h_adjusted[central_adjusted] = np.minimum(lower_dist[central_adjusted],
+                                              upper_dist[central_adjusted])
+    return h_adjusted, scheme
+
+
+def _approx_forward(f, x0, h, f0, args, bounds):
+    h = _adjust_step_1st_order(x0, h, bounds)
+
+    if f0 is None:
+        f0 = f(x0, *args)
+
+    if x0.ndim == 0:
+        x = x0 + h
+        dx = x - x0
+        df = f(x, *args) - f0
+        J = df / dx
+    else:
+        J = []
+        for i in range(x0.size):
+            x = np.copy(x0)
+            x[i] += h[i]
+            dx = x[i] - x0[i]
+            df = f(x, *args) - f0
+            J.append(df / dx)
+        J = np.array(J).T
+    return J
+
+
+def _approx_central(f, x0, h, f0, args, bounds):
+    h, scheme = _adjust_step_2nd_order(x0, h, bounds)
+
+    if np.any(scheme == 1) and f0 is None:
+        f0 = f(x0, *args)
+
+    if x0.ndim == 0:
+        if scheme == 0:
+            x1 = x0 - h
+            x2 = x0 + h
+            dx = x2 - x1
+            f1 = f(x1, *args)
+            f2 = f(x2, *args)
+            J = (f2 - f1) / dx
+        else:
+            x1 = x0 + h
+            x2 = x0 + 2 * h
+            dx = x2 - x0
+            f1 = f(x1, *args)
+            f2 = f(x2 * h, *args)
+            J = (-3 * f0 + 4 * f1 - f2) / dx
+    else:
+        J = []
+        for i in range(x0.size):
+            x1 = np.copy(x0)
+            x2 = np.copy(x0)
+            if scheme[i] == 0:
+                x1[i] -= h[i]
+                x2[i] += h[i]
+                dx = x2[i] - x1[i]
+                f1 = f(x1, *args)
+                f2 = f(x2, *args)
+                J.append((f2 - f1) / dx)
+            else:
+                x1[i] += h[i]
+                x2[i] += 2 * h[i]
+                f1 = f(x1, *args)
+                f2 = f(x2, *args)
+                dx = x2[i] - x0[i]
+                J.append((-3 * f0 + 4 * f1 - f2) / dx)
+        J = np.array(J).T
+
+    return J
+
+
+def approx_derivatives(f, x0, method='central', rel_step=None,
+                       f0=None, args=(), bounds=(None, None)):
```

Comment:
Let's make the name a singular `approx_derivative`, not plural. Rationale: it computes a single object (derivative, gradient, Jacobian) rather than, say, first and second derivatives at a given point.

Reply:
Ever considered to use the Python `autograd` package as the default method for Jacobian calculation? It is much faster than `numdifftools`